### ISSUE TITLE
Worlds and singleton tag components

### ISSUE THREAD

**Issue by player-03** [onehundredfeet/hmecs#3]
I haven't tested this, but I'm struggling to see what benefit world flags provide over simple components. I mean, isn't this...

```haxe
static function main() {
  //...
  var john = createRabbit(0, 0, 1, 1, 'John', WORLDS_FIELDS); // Only in the forest
  var jack = createRabbit(5, 5, 1, 1, 'Jack', WORLDS_FIELDS | WORLDS_FOREST); // In both worlds
  //...
}

//Can narrow the scope of the update to only entities that are present in a world set
@:worlds(WORLDS_FOREST) // These are bit flags.  The string is evaulate as an expression
@update function inForest(name:Name) {
  trace('${name} is in the forest'); // Will display Jack
}

// Worlds can be strings or constant string expressions if using compiler only @:
@:worlds(WORLDS_FIELDS) // These are bit flags.  The string is evaulate as an expression
@:update function inFields(name:Name) {
  trace('${name} is in the fields'); // Will display Jack & John
}
```

...functionally the same as this?

```haxe
static function main() {
  //...
  var john = createRabbit(0, 0, 1, 1, 'John').add(FieldsWorld.instance); // Only in the forest
  var jack = createRabbit(5, 5, 1, 1, 'Jack').add(FieldsWorld.instance, ForestWorld.instance); // In both worlds
  //...
}

@update function inForest(name:Name, world:ForestWorld) {
  trace('${name} is in the forest'); // Will display Jack
}

@:update function inFields(name:Name, world:FieldsWorld) {
  trace('${name} is in the fields'); // Will display Jack & John
}
```

**Comment by onehundredfeet** [comment]
To be honest, I think you may be right. I implemented worlds as the first feature to reduce the overhead of using the components, but I have since implemented a flag component that is used a highly compressed and non-undirected representation that effectively shadows worlds.

The only use case I can see is world global components.  I bet that can be worked around too.

I'll think about it for a bit longer and remove them if I can't find anything that requires them.

**Comment by player-03** [comment]
I was always unclear on the purpose of global and singleton components. If components aren't part of an entity, why do they have to be components? Are they even components? Why not just use a normal singleton?

That said, I do have an idea for world global components.

```haxe
static function main() {
  //...
  FieldsWorld.instance = new Entity();
  FieldsWorld.instance.add((0x99FF99:WorldBackgroundColor));
  FieldsWorld.instance.add(("fields":WorldName));
  
  ForestWorld.instance = new Entity();
  ForestWorld.instance.add((0x00CC00:WorldBackgroundColor));
  ForestWorld.instance.add(("forest":WorldName));
  
  var john = createRabbit(0, 0, 1, 1, 'John').add(FieldsWorld.instance); // Only in the forest
  var jack = createRabbit(5, 5, 1, 1, 'Jack').add(FieldsWorld.instance, ForestWorld.instance); // In both worlds
  //...
}

@update function inForest(name:Name, world:ForestWorld) {
  trace('${name} is in the ${ForestWorld.instance.get(WorldName)}'); // Will display Jack
}

@:update function inFields(name:Name, world:FieldsWorld) {
  trace('${name} is in the ${FieldsWorld.instance.get(WorldName)}'); // Will display Jack & John
}
```

That said, the value of the `world` argument is entirely useless. It can only be `1` or `0` (question: why not `true`/`false`?), and if the function is being called it can only be `1`.

**Comment by player-03** [comment]
Here's an idea: tag components can also be singletons. All you have to do is define the value in advance:

```haxe
package world;

@:storage(TAG) @:tagInst(world.ForestWorld.instance) @:forward
abstract ForestWorld(Entity) {
	public static var instance(default, null):ForestWorld = cast new Entity();
}
```

Then in `getGetExpr()`, instead of returning `? 1 : 0`, you look up the provided expression.

```haxe
public function new(ct:ComplexType, i:Int, pos) {
	//...
	storageType = StorageType.getStorageType(followedMeta);
	instExpr = macro true;
	if (storageType == TAG) {
		var tagMeta = followedMeta.get(":tagInst");
		if (tagMeta != null && tagMeta[0] != null) instExpr = tagMeta[0];
	}
	//...
}

//...

public function getGetExpr(entityExpr:Expr, cachedVarName:String = null):Expr {
	var container = cachedVarName != null ? macro $i{cachedVarName} : macro $containerFullNameExpr.storage;
	return switch (storageType) {
		case FAST: macro $container[$entityExpr];
		case COMPACT: macro $container.get($entityExpr);
		case SINGLETON: macro $container;
		case TAG: var te = tagExpr();
		  macro @:privateAccess ecs.Workflow.getTag($entityExpr, $te) ? $instExpr : null;
	};
}
```

---

With this change, the arguments would finally have a useful value.

```haxe
@:update function inForest(name:Name, world:ForestWorld) {
  trace('${name} is in the ${world.get(WorldName)}'); // Will display Jack
}

@:update function inFields(name:Name, world:FieldsWorld) {
  trace('${name} is in the ${world.get(WorldName)}'); // Will display Jack & John
}
```

**Comment by onehundredfeet** [comment]
That's a really interesting idea.  Let me process for a bit.

**Comment by onehundredfeet** [comment]
I really like the core of this idea, of saying a tag can be associated with a singleton, although, that's pretty much exactly what the current singleton does without any additional markup.

However, your method above would make it so a new class wouldn't have to be created each time.

It also alludes to SYSTEM level scoping of bound singletons, which is VERY interesting.  i.e. the same tag could lead to different singletons. Which ultimately leads back to where you started which is that they are just system level variables.

I think the case for ECS level singletons is that the singleton is naturally global vs systems.  I.e. it's a capability that is exposed through the ECS but available to all systems.

System level globals can just be members are you show.

**Comment by player-03** [comment]
> that's pretty much exactly what the current singleton does without any additional markup.

Not if you throw an error the moment you try to add the singleton to a second entity. I do see why you throw that error: otherwise, calling `entityB.add(new SingletonComponent())` could change the value on `entityA`.

The point of the extra markup is to get around that. There's one canonical spot to define the singleton, and then you can call `entity.add()` as many times as you like without overwriting it.

> It also alludes to SYSTEM level scoping of bound singletons, which is VERY interesting. i.e. the same tag could lead to different singletons. Which ultimately leads back to where you started which is that they are just system level variables.

Not sure how it alludes to that. Yeah maybe you could pull that off, but what's the upside? The downside is pretty obvious: it's more efficient to hard-code the instance into the `get` function.

I don't think I ever suggested that singletons should be system-level variables. Static variables, sure, but my suggestion was to initialize them in `main()` and store them in their own files. (`ForestWorld.instance` in ForestWorld.hx, etc.) As a user, I would never want to have to redefine all my singletons each time I made a new system.

> I think the case for ECS level singletons is that the singleton is naturally global vs systems. I.e. it's a capability that is exposed through the ECS but available to all systems.

Not sure how to parse this, but I looked at the code and came up with some things.

Since `getGetExpr()` lacks any ownership checks, you can call `entityA.add(new SingletonComponent())` and then get it via `entityB.get(SingletonComponent)`. At first I thought that was meant for convenience, but now I'm pretty sure it's just a bug.

Between `getAddExpr()` and README.md, it's clear that only one entity is supposed to have the singleton component. (README.md also claims that only one instance of `SingletonComponent` may exist, but the constructor is public so that simply isn't true.)

Are there benefits to the one-entity limit? Perhaps. You can access the singleton entity via system functions, with all the cool features that come with that. You get the `SingletonComponent` instance, plus any number of associated components. And you can be confident that the system functions will only be called once (or only once per update).

But are those benefits? Gotta be honest: I don't see any use cases that aren't better handled some other way.

- If I wanted access to the `SingletonComponent` instance, I'd make it a proper singleton and then access `SingletonComponent.instance`.
- If I wanted a unique entity whose components had a special meaning, I'd make a normal entity and then store it as `Main.mySpecialEntity`. And I could do this multiple times if I had multiple special meanings in mind.
- If I wanted a component that could only be added to one entity... actually I can't conceive of wanting that. But if I did, I'd make an `@:add` listener that threw an error the second time.

I tried to look up some use cases online, but [the few I found](https://gamedev.stackexchange.com/q/77883/130579) don't actually need to be singleton components. `World` should not be a singleton - what if you ever decide to preload the next level in the background? `GameState` is in the same boat. What if you want to make split-screen multiplayer, where each player's half of the screen has its own game state? `Keyboard` probably would be a singleton, but there's no need for it to be a component. `Keyboard.instance` is simpler.

My feeling is, the ECS pattern is at its best when it doesn't set limits. You can have one `GameState`, or you can have five, and systems won't care. As long as an entity has the right components, systems will know what to do with it. And if a task really does need to happen only once... that's what static functions are for.

**Comment by onehundredfeet** [comment]
It seems to have eaten my reply.  I'll post it again tomorrow.

Btw - I really appreciate the engagement to this depth.  It doesn't happen often.

**Comment by onehundredfeet** [comment]
Okay, my initial response was different than what I'm going to say here after a night to think about it.  I think your idea about using tags with statics is really elegant.  Here's a first pass, but I think it can get simpler. My initial Tag design was rushed.  Tags are already in the system, but with this change, maybe it can be simpler.

What about this:
```haxe
//API Proposal: Extending the current API by associating static expressions with tags
// This would configure the tag to pass the expression 'MyClass.instance' instead of 0|1 into the system's function
// Evolution of current design, but enhanced with the substitution of expression vs simply a 'true|false' parameter.
@:storage(TAG, MyClass.instance)
@:enum abstract TagB(Int) from Int to Int {
    var INVALID = 0;
    var VALID = 1;
}

function addTagExample(e:Entity) {
   //Current API
  e.add(TagB.VALID);

  //Proposed API cleanup
  e.tag(TagB);
}

@:update
function myUpdate( tag : TagB) {
   // Tag would be substitute for MyClass.instance, which I think is sketchy.
}

//
// API Idea 2 - Class to tag is 1:1, still uses types as component keys
//
@:storage(TAG) // implicitly adds a static 'instance' of MyTag that is passed in 
class MyTag {
   public MyTag() {}

    @:first  //optional
   static function calledWhenFirstEntityIsTagged() {}
   
   @:last  //optional
   static function calledWhenLastTaggedEntityIsUntagged() {}
   
   @:factory  //optional - Called when static instance is needed
   static MyTag makeMyTag() return new MyTag();
}

function addTagExample2(e:Entity) {
  //Proposed API
  //Would automatically create the instance if needed and call callbacks
  e.tag(MyTag);
}

@:update
function myUpdate( tag : MyTag) {
      // Tag would be the static instance
}

//If you don't need the variable...
@:update(MyTag) // updates only with these tags
function extension( ) {
  // This would avoid the overhead of reading the variable if it's not needed
}
```


ok, done.

**Comment by player-03** [comment]
Thinking about this myself, I think there's a good case to be made for merging `TAG` with `SINGLETON`, since they're already so close conceptually. Both are limited to two values: `1` or `0` for `TAG`, the singleton or `null` for `SINGLETON`, and both take advantage of that fact to achieve efficient storage.

The question is how much of each to keep. And, I guess which name to call it.

Setting the value via `entity.add()` (as with `SINGLETON`) is the API that users are used to. However, without the "you can only add it once" restriction, it becomes very easy to overwrite the value. My approach to `TAG` is designed to solve this problem, but has the downside that it isn't intuitive. As a third option, you could try a function like `Workflow.setTagValue()`.

Some users do want and expect the behavior where you can only add a component to one entity. It wasn't just you - I found multiple other instances of this. I've already gone over my objections to this feature, but I realize mine isn't the only perspective. At the very least, I should probably explain why I omitted the feature in my fork, since people would otherwise ask.

> ```haxe
> //Proposed API
> //Would automatically create the instance if needed and call callbacks
> e.tag(MyTag);
> ```

Something like that could certainly work. The only information you actually need to pass is the `Class<T>`, and hmecs can look up the rest. Passing the instance that was also defined by metadata is redundant, and passing a different instance is incorrect.

The other approach I'd consider is straight-up `e.add(MyTag)`. This would make the `add()` function more complicated, but keep the API more accessible.

Actually the `add()` function is already that complicated, and might need to become more complicated. You aren't supposed to call `e.add(new MyTag())`, so it ought to check for that and throw an error. So what's a little more complication on top of that?

> ```haxe
> // API Idea 2 - Class to tag is 1:1, still uses types as component keys
> ```

Yeah, that's definitely required, otherwise writing system functions becomes much harder. Or "sketchy," as you put it. The argument type is `TagB` but its value is `MyClass.instance`? That won't do.

> ```haxe
> @:first  //optional
> static void calledWhenFirstEntityIsTagged() {}
> 
> @:last  //optional
> static void calledWhenLastTaggedEntityIsUntagged() {}
> ```

I've seen component-level add/remove listeners before, and always thought they could just as easily go in systems. The "first" and "last" approach is new, but I'm pretty sure that too could go in a system. It'd be a pair of listeners in the `View` class, for when the number of entities in view went from 0 to 1, or 1 to 0. Bonus: that could apply either to views of a single component, or views of multiple.

The one thing I will say is that, if you go ahead with this feature, there are zero reasons to limit it to tags. Whether you do it via `View` or in the component class, all component types would benefit (or not) equally.

That said, I also don't see the use case for `@:first` and `@:last`. My approach to ECS is that the first entity should work the same as the tenth as the thousandth. If I had to know whether an entity was first, I could check `view.entities.length`, but first I would stop and think that maybe I'm writing the wrong code.

> ```haxe
> @:factory  //optional - Called when static instance is needed
> static MyTag makeMyTag() return new MyTag();
> ```

Huh. I was going to say that this wouldn't work, but I think I'm misremembering. I've kind of internalized that you can't examine a class's functions from anything other than a build macro for that class, and since there was no `@:build` metadata, I figured you wouldn't be able to get it. But no, the thing you can't access is the function `Expr`, the `return new MyTag()` part. You should be able to access the metadata and function name, which ought to be all you need.

Anyway, that tangent aside, this can certainly work, though again it's worth considering the `Workflow.setTagInstance()` approach. The latter might be easier to explain to users.

**Comment by onehundredfeet** [comment]
> The other approach I'd consider is straight-up `e.add(MyTag)`. This would make the `add()` function more complicated, but keep the API more accessible.
> 
> Actually the `add()` function is already that complicated, and might need to become more complicated. You aren't supposed to call `e.add(new MyTag())`, so it ought to check for that and throw an error. So what's a little more complication on top of that?

Agreed.  


> The one thing I will say is that, if you go ahead with this feature, there are zero reasons to limit it to tags.

Agreed.

> Some users do want and expect the behavior where you can only add a component to one entity. It wasn't just you - I found multiple other instances of this. I've already gone over my objections to this feature, but I realize mine isn't the only perspective. At the very least, I should probably explain why I omitted the feature in my fork, since people would otherwise ask.

Is it acceptable to add metadata for the components you want to use that way?

```haxe
@:sharable
class MyComponent {

}
```

**Comment by player-03** [comment]
Sharable? Not sure what you're getting at. I was talking about the `SINGLETON` storage type.

**Comment by onehundredfeet** [comment]
> Sharable? Not sure what you're getting at. I was talking about the "singleton component" behavior.

Ahh..

Well the above tag design would allow users to add the tag to as many entities as you want and they would all access the same static instance.

In addition, right now, I'm not sure there's any restrictions for adding the same component to multiple entities, other than any ones that use object pools.

**Comment by onehundredfeet** [comment]
Quick sharable modification for non-singletons

```haxe
#if !macro @:build(ecs.core.macro.PoolBuilder.arrayPool(true)) #end // make this sharable
class SmallComponent {
}
```

or an alternate
```haxe
#if !macro @:build(ecs.core.macro.PoolBuilder.arrayPool()) #end // make this sharable
@:sharable
class SmallComponent {
}
```


This would refcount the objects so that it doesn't free multiple times.

**Comment by player-03** [comment]
Yeah, that sounds like a pool-specific issue. I've never had any problems with reusing components otherwise. Since it's related to pools, I say pass the argument to `PoolBuilder`.

**Comment by onehundredfeet** [comment]
Ok, I just did an initial implementation of the flag / singleton hybrid.

It's pretty sweet:

```haxe
@:storage(TAG)
class TagA {
    public function new () {}
    public var test = 0;
}

function run() {
   var e1 = new Entity();
   var e2 = new Entity();
   
   e1.add( TagA );
   e1.get(TagA).test = 1;
   
   e2.add(TagA);
   trace('${e2.get(TagA).test == 1}'); // true
}
```

**Comment by onehundredfeet** [comment]
I can check it in, but I should probably do a bit more testing first.

It's a breaking change.

**Comment by player-03** [comment]
So... looks you call `add()` like any other component, and it calls `new TagA()`? That means that you have to provide a constructor, and the constructor can't have required arguments. That's fine in many cases, and certainly makes the code easy to understand, but it'll put some limits on what can be used as a `TAG` component.

It's probably fine. At least, off the top of my head I can't come up with a reason why a singleton instance would need arguments. Arguments help differentiate between multiple instances, and if there's only meant to be one instance, you can hard-code any relevant information.

Though again, I'd recommend making the constructor private. Or more to the point, I think users might want to write private constructors, to remind themselves not to call it. So just make sure you include `@:privateAccess` when calling the constructor, if you didn't already.

**Comment by onehundredfeet** [comment]
Agreed on all points.

**Comment by onehundredfeet** [comment]
Posted commit

**Comment by player-03** [comment]
I don't see why it has to be a class; abstracts have constructors too.

Also, `getGetExpr()` is now incorrect. If it receives `cachedVarName`, it will now return the tag instance, whether or not this entity is tagged. (I suggest simply removing the redundant `switch` block, as shown [here](https://github.com/onehundredfeet/hmecs/pull/4/files#diff-4d84fc8f157214e15bd3b47c010a21f5560b9ad4d5ee6c9235baaebd9df4295eL202-L219).)

**Comment by onehundredfeet** [comment]
That's a bit of me being lazy. I should really separate them into two functions.  The only use case of the cached flow is through view iteration, which already knows that all the entities have the tag, so it doesn't need to check.  I should probably move it into another function.

**Comment by onehundredfeet** [comment]
I just checked in the un-lazy version.  If there's something I'm still missing let me know.

**Comment by onehundredfeet** [comment]
﻿On the whole I agree with you.  I often preach the same principles you are below.  I think that I was trying to provide coverage in the case there was a case that emerged.  Due to the way the components are stored, using a singleton provides some amount of advantage over the compact map representation, but the extra complexity may not be worth it.  

In my game I a server, client sim and presentation worlds.  Each had a system entity that held any global components for each world.  While I could have made those components static somewhere I felt it betrayed the exposed state philosophy of ECS. For all the reasons you point out, I’m not sure if that pattern is good.  

On the other hand, The entity with a singleton does provided all the other lifetime management features such as signals when added and removed, without the need for extra logic per external singleton.

I.e. class.instance may require additional work to initialize and any dependant processing then requires extra infrastructure.  You get that for free when embedded in the ecs.

**Comment by player-03** [comment]
> lifetime management features such as signals when added and removed

That hadn't really occurred to me. Probably because in my code, singletons are created the first time anything tries to access them, and then never destroyed. If at any point a singleton became null, I'd consider that an error.

Do you have an example of when you'd add/remove a singleton?

> class.instance may require additional work to initialize [...] You get that for free when embedded in the ecs.

It's pretty easy to automate either way. The real question is where to put the infrastructure, and my answer is "[in its own tiny Haxelib](https://lib.haxe.org/p/SingletonInterface)."

**Comment by player-03** [comment]
I've been thinking about worlds again recently, and since this thread is at least partially about them, I might as well post here. (Though in retrospect I wish I'd made two issues for the two concepts.)

If someone came from another ECS looking for a "worlds" feature, they would probably expect to be able to create entirely separate simulations. They'd want to be able to create and destroy these simulations programmatically, and they'd want each simulation to have its own independent list of views and systems. That way if they made a huge simulation with hundreds of systems, it wouldn't bog down the others.

Implementing this would mainly involve turning singletons into arrays. Instead of a singleton of each `View`, you'd have an array of views, indexed by world ID. Same for `ViewsOfComponent` and `Workflow.systems`. Each world would have its own `Array<Bool>` to track which entities belong to that world. Entities could belong to any number of worlds at a time, and they'd bring all their components with them (component storage would still use singletons).

**Comment by onehundredfeet** [comment]
I think that with regards to the singletons and tags, this has been addressed.
